refactor: Create BlockService with block related classes

konstantinabl opened the pull request:
**Description**:
<!--
One or two line summary of what this PR does and why it is needed, followed by a list
of changes in imperative, present tense for use in the commit message or changelog. Example:

Add support for ...

* Add config property
* Change column name
* Remove ...
-->

This PR refactors the JSON-RPC relay by extracting block-related functionality from the eth.ts file into a dedicated BlockService class and moving shared code into a CommonService. This improves code organization, testability, and maintainability.

**Changes** 

- Created new BlockService to handle all block-related RPC methods
- Moved methods used only from the block related methods to the blockService - **_populateSyntheticTransactions, getTransactionCountFromBlockResponse_**
- Extracted common functionality into a CommonService class - **_formatContractResult_**
- Fixed type issues in getBlock method
- Made some calls to MN in getBlock method parallel to improve performance
- Added JSDoc where missing

**Moved methods moved to BlockService**:

Main methods:

- getBlockByNumber
- getBlockByHash
- getBlockTransactionCountByHash
- getBlockTransactionCountByNumber
- getBlockReceipts
- getUncleByBlockHashAndIndex
- getUncleByBlockNumberAndIndex
- getUncleCountByBlockHash
- getUncleCountByBlockNumber

Utility methods:

- getBlock
- getTransactionCountFromBlockResponse
- populateSyntheticTransactions

**Related issue(s)**:

Fixes #3569

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

Ferparishuertas commented on packages/relay/src/lib/services/blockService/IBlockService.ts:
```diff
@@ -10,6 +10,12 @@ export interface IBlockService {
+  getUncleCountByBlockNumber: (requestDetails: RequestDetails) => Promise<string>;
 }
 
 export interface ITimestamp {
```
Why do we add more than 1 interface in the same file?

konstantinabl replied:
Thanks for pointing this out, this is not my change actually and I didnt see it. It comes from an upstream branch which has Niki's changes. I will fix it

konstantinabl replied:
Fixed now

Ferparishuertas commented on packages/relay/src/lib/services/blockService/index.ts:
```diff
@@ -19,19 +19,57 @@ import { TransactionFactory } from '../factories/transactionFactory';
 import { IBlockMirrorNode, IBlockService } from './IBlockService';
 
 export class BlockService implements IBlockService {
```
Why are we calling this file index.ts when it is not really a module, and visibility is across the relay?

konstantinabl replied:
Fixed it : )

Ferparishuertas commented on packages/relay/src/lib/services/ethService/ethCommonService/index.ts:
```diff
@@ -16,20 +17,19 @@ import {
-import HAPIService from '../../hapiService/hapiService';
 import { TransactionFactory } from '../../factories/transactionFactory';
+import HAPIService from '../../hapiService/hapiService';
+import { ICommonService } from './ICommonService';
```
Why are we calling this file index.ts when it is not really a module, and visibility is across the relay?

and why the package is ethervice/ethCommonService?
Why not simply CommonService?
If we want to distinguish eth services why not. eth.services, and the block, common and others?

konstantinabl replied:
Changed the name, its CommonService.ts now

Ferparishuertas commented on packages/relay/src/formatters.ts:
```diff
@@ -6,6 +6,7 @@ import { BigNumber as BN } from 'bignumber.js';
 
 import constants from './lib/constants';
+import { Transaction } from './lib/model';
 import { TransactionFactory } from './lib/services/factories/transactionFactory';
```
Why this file is at the main level, and why is it lowerCase and others Uppercase?

konstantinabl replied:
Not sure what you mean by the "main level"

Ferparishuertas replied:
I mean at src level. We wiil review it outside of the PR

Ferparishuertas commented:
Can you enumerate which methods are extracted and why?
I do see missing getBlockReceiptsm and other private methods, that are block related

Ferparishuertas commented on packages/relay/src/lib/eth.ts:
```diff
@@ -25,7 +25,6 @@ import {
 } from '../formatters';
```
Why eth is inside a lib package?

simzzz commented on packages/relay/src/lib/services/blockService/BlockService.ts:
```diff
@@ -0,0 +1,424 @@
+    this.cacheService = cacheService;
+  }
+
+  public async getBlockByNumber(
```
Since we have separate handlers for `getBlockByNumber` and `getBlockByHash` in eth.ts wouldn't it be cleaner to have one function for both here since they are very similar in order to reduce code repetition and just pass some extra properties. 

An example signature could be
```
private async getBlockByNumberOrHash(
  type: "number" | "hash",
  identifier: string,
  showDetails: boolean,
  requestDetails: RequestDetails,
  cacheKeyPrefix: string,
  cacheMethod: Function,
): Promise<Block | null> 
```

konstantinabl replied:
Will resolve this in a separate issue, so we can discuss the logging

simzzz commented on packages/relay/src/lib/services/blockService/BlockService.ts:
```diff
@@ -0,0 +1,424 @@
+    this.logger.trace(`${requestIdPrefix} getBlockByNumber(blockNumber=${blockNumber}, showDetails=${showDetails})`);
+
+    const cacheKey = `${constants.CACHE_KEY.ETH_GET_BLOCK_BY_NUMBER}_${blockNumber}_${showDetails}`;
+    let block = await this.cacheService.getAsync(cacheKey, EthImpl.ethGetBlockByNumber, requestDetails);
```
Shouldn't `EthImpl.ethGetBlockByNumber` be moved to BlockService along with other statics like in `getBlockByHash`?

konstantinabl replied:
fixed

natanasow commented on packages/relay/src/lib/sth.json:
```diff
@@ -0,0 +1,541 @@
+//contract results
```
q: what's the purpose of these files `sth.json` and `sth2.json`, some comment might be useful

konstantinabl replied:
accidentally added. removed now : )

konstantinabl commented on packages/relay/src/lib/services/blockService/BlockService.ts:
```diff
@@ -0,0 +1,532 @@
+      const to = await this.common.resolveEvmAddress(contractResult.to, requestDetails);
+
+      const contractAddress = this.common.getContractAddressFromReceipt(contractResult);
+      const receipt = {
```
Yes, I was thinking the same, I will add an issue to the refactor epic

konstantinabl replied:
https://github.com/hiero-ledger/hiero-json-rpc-relay/issues/3622

konstantinabl commented on packages/relay/src/lib/services/blockService/BlockService.ts:
```diff
@@ -0,0 +1,532 @@
+    mirrorNodeClient: MirrorNodeClient,
+    logger: Logger,
+  ) {
+    this.common = common;
```
fixed.

konstantinabl commented on packages/relay/src/lib/eth.ts:
```diff
@@ -2132,9 +2031,9 @@ export class EthImpl implements Eth {
+    const block = await this.mirrorNodeClient.getBlock(blockHash, requestDetails);
     if (block != null) {
-      return block.number;
+      return numberTo0x(block.number);
```
https://github.com/hiero-ledger/hiero-json-rpc-relay/pull/3600/files/81982e90829842889d034f9a3beab6ce63aa0fb3#diff-dd8c7990f7764960d05dcb4c1b5ae7eee48b425b1b83d6dcc7e7ccc442330ef0R2034

Because of the change on line 2034 above. We dont call the whole getBlockByHash flow anymore, just the mirrorNode, so the MN response contains the block.number, but not in hex format

konstantinabl commented:
> Can you enumerate which methods are extracted and why? I do see missing getBlockReceiptsm and other private methods, that are block related

@Ferparishuertas I have moved getBlockReceipts now and listed the methods i have moved. Not sure how to explain why, without making it too long. I moved the block related methods to BlockService and any methods used by those methods to the blockService as well. Also moved any utility methods used by the block-related methods and other methods from eth.ts to CommonService, so eth.ts doesn't include any business logic and utility methods. 
What other private methods are missing?

Ferparishuertas reviewed (approved):
LGTM
